src/mono/wasm/wasm.proj

```diff
@@ -89,6 +80,82 @@
   </Target>
 
+  <UsingTask TaskName="Microsoft.WebAssembly.Build.Tasks.EmitWasmBundleSourceFiles" AssemblyFile="$(WasmAppBuilderTasksAssemblyPath)" />
+  <Target Name="GenerateTimezonesArchive" Returns="@(_WasmArchivedTimezones)">
+    <PropertyGroup>
+      <_WasmTimezonesPath>$([MSBuild]::NormalizePath('$(PkgSystem_Runtime_TimeZoneData)', 'contentFiles', 'any', 'any', 'data'))</_WasmTimezonesPath>
+      <_WasmTimezonesBundleSourceFile>$(WasmObjDir)\wasm-bundled-timezones.c</_WasmTimezonesBundleSourceFile>
+      <_WasmTimezonesBundleObjectFile>$(WasmObjDir)\wasm-bundled-timezones.o</_WasmTimezonesBundleObjectFile>
+      <_WasmTimezonesBundleArchive>$(WasmObjDir)\wasm-bundled-timezones.a</_WasmTimezonesBundleArchive>
+      <_WasmTimezonesSourcesRsp>$(WasmObjDir)\wasm-bundled-timezones-sources.rsp</_WasmTimezonesSourcesRsp>
+      <_WasmTimezonesArchiveRsp>$(WasmObjDir)\wasm-bundled-timezones-archive.rsp</_WasmTimezonesArchiveRsp>
+    </PropertyGroup>
+    <ItemGroup>
+      <_WasmTimezonesInternal Include="$(_WasmTimezonesPath)\**\*.*" WasmRole="Timezone"/>
+    </ItemGroup>
+    <GetFileHash Files="@(_WasmTimezonesInternal)">
+      <Output TaskParameter="Items" ItemName="_WasmBundleTimezonesWithHashes" />
+    </GetFileHash>
+    <ItemGroup>
+      <_WasmBundleTimezonesWithHashes Update="@(_WasmBundleTimezonesWithHashes)">
+        <DestinationFile>$(WasmObjDir)\wasm-bundled-$([System.String]::Copy(%(_WasmBundleTimezonesWithHashes.FileHash))).c</DestinationFile>
+        <ObjectFile>$(WasmObjDir)\wasm-bundled-$([System.String]::Copy(%(_WasmBundleTimezonesWithHashes.FileHash))).o</ObjectFile>
+        <RegisteredName>/usr/share/zoneinfo/$([MSBuild]::MakeRelative($(_WasmTimezonesPath), %(_WasmBundleTimezonesWithHashes.Identity)).Replace('\','/'))</RegisteredName>
+      </_WasmBundleTimezonesWithHashes>
+    </ItemGroup>
+    <!-- TODO make this incremental compilation -->
+    <EmitWasmBundleSourceFiles
+      FilesToBundle="@(_WasmBundleTimezonesWithHashes)"
+      BundleName="timezones"
+      BundleFile="$(WasmObjDir)\wasm-bundled-timezones.c"
+      RegistrationCallbackFunctionName="mono_wasm_add_bundled_file"
+      >
+    </EmitWasmBundleSourceFiles>
+    <ItemGroup>
+      <_WasmBundleTimezonesSources Include="$([MSBuild]::MakeRelative($(WasmObjDir), %(_WasmBundleTimezonesWithHashes.DestinationFile)).Replace('\','/'))" />
+      <_WasmBundleTimezonesSources Include="$([MSBuild]::MakeRelative($(WasmObjDir), $(_WasmTimezonesBundleSourceFile)).Replace('\','/'))" />
+    </ItemGroup>
+    <WriteLinesToFile File="$(_WasmTimezonesSourcesRsp)"
+      Overwrite="true"
+      Lines="@(_WasmBundleTimezonesSources, ' ')"
+      WriteOnlyWhenDifferent="true" />
+    <RunWithEmSdkEnv Command="$(EmccCmd) -xc -c @$(_WasmTimezonesSourcesRsp)"
+          WorkingDirectory="$(WasmObjDir)"
+          EmSdkPath="$(EMSDK_PATH)"
+          ConsoleToMsBuild="true"
+          IgnoreStandardErrorWarningFormat="true">
+    </RunWithEmSdkEnv>
```

Comment:
For a follow up PR, consider using `EmccCompile` task which compiles in parallel.